components/tidb_query/src/codec/chunk/chunk.rs

```diff
@@ -102,14 +102,28 @@ impl Chunk {
                 }
             }
             VectorValue::Real(ref vec) => {
-                for row_index in row_indexs {
-                    let row_index = *row_index as usize;
-                    match &vec[row_index] {
-                        None => {
-                            col.append_null().unwrap();
+                if col.get_fixed_len() == 4 {
+                    for row_index in row_indexs {
+                        let row_index = *row_index as usize;
+                        match &vec[row_index] {
+                            None => {
+                                col.append_null().unwrap();
+                            }
+                            Some(val) => {
+                                col.append_f32(f64::from(*val) as f32).unwrap();
```

Comment:
I remember TiKV doesn't do anything about float32 overflow, does it? @breeswish

Reply:
Ok, I got it. F32 only comes from constant and column data so it never overflows. Therefore, it's safe to cast it from f64 to f32.